src/julia_threads.h

```diff
@@ -330,9 +333,9 @@ STATIC_INLINE int8_t jl_gc_state_set(jl_ptls_t ptls, int8_t state,
                                      int8_t old_state)
 {
     jl_atomic_store_release(&ptls->gc_state, state);
-    if (state == JL_GC_STATE_SAFE && old_state == 0)
+    if ((state == JL_GC_STATE_SAFE) && old_state == JL_GC_STATE_UNSAFE)
         jl_gc_safepoint_(ptls);
-    if (state == 0 && old_state == JL_GC_STATE_SAFE)
+    if (state == JL_GC_STATE_UNSAFE && (old_state == JL_GC_STATE_SAFE || old_state == JL_GC_STATE_WAITING))
         jl_gc_safepoint_(ptls);
```

Comment:
We could also simplify this to:
```suggestion
    if (state == JL_GC_STATE_UNSAFE || old_state == JL_GC_STATE_UNSAFE)
        jl_gc_safepoint_(ptls);
```
While not exactly equivalent to your version, I think they cover identically all of the transitions that actually occur.

In particular, the two versions would differ on the transition
old_state=JL_GC_STATE_UNSAFE, state=JL_GC_STATE_UNSAFE
where my version does an unnecessary (faster) safepoint instead of (a slower) branching to skip it

and also on the transition
old_state=JL_GC_STATE_UNSAFE, state=JL_GC_STATE_WAITING
which doesn't normally occur, but should also is supposed to be fine to have a safepoint there too